libs/common/src/helpers/date-validation.ts

```diff
@@ -0,0 +1,130 @@
+      return {
+        errRefs: [valueRef],
+        errMsg: `Enter the ${valueName}`,
+      };
+    }
+    if (!inputtedDay && !inputtedMonth) {
+      return {
+        errRefs: [dayFieldRef, monthFieldRef],
+        errMsg: `${valueName} must include a day and a month`,
+      };
+    }
+    if (!inputtedDay && !inputtedYear) {
+      return {
+        errRefs: [dayFieldRef, yearFieldRef],
+        errMsg: `${valueName} must include a day and a year`,
+      };
+    }
+    if (!inputtedMonth && !inputtedYear) {
+      return {
+        errRefs: [monthFieldRef, yearFieldRef],
+        errMsg: `${valueName} must include a month and a year`,
+      };
+    }
+    if (!inputtedDay)
+      return {
+        errRefs: [dayFieldRef],
+        errMsg: `${valueName} must include a day`,
+      };
+  }
+  if (!inputtedMonth) {
+    return {
+      errRefs: [monthFieldRef],
+      errMsg: `${valueName} must include a month`,
+    };
+  }
+  if (!inputtedYear) {
+    return {
+      errRefs: [yearFieldRef],
+      errMsg: `${valueName} must include a year`,
+    };
+  }
+
+  // schema to validate that the year is 4 digits long and only numbers
+  const yearFormatSchema = Joi.string().length(4).pattern(/^\d+$/).required();
+  const yearFormatValidation = yearFormatSchema.validate(inputtedYear.toString());
+
+  // schema which ensures that the month is an integer between 1 and 12 (allows for leading 0s)
+  const monthSchema = Joi.string().pattern(/^([1-9]|1[012]|0[1-9])$/);
+  const monthFormatValidation = monthSchema.validate(inputtedMonth.toString());
+
+  // schema which ensures that the day is an integer between 1 and 31 (allows for leading 0s)
+  const daySchema = Joi.string().pattern(/^(3[01]|[12][0-9]|[1-9]|0[1-9])$/);
+  const dayFormatValidation = daySchema.validate(inputtedDay.toString());
+
+  if (dayFormatValidation.error || monthFormatValidation.error || yearFormatValidation.error) {
+    if (
+      (dayFormatValidation.error && (monthFormatValidation.error || yearFormatValidation.error)) ||
+      (monthFormatValidation.error && yearFormatValidation.error)
+    ) {
+      return {
+        errRefs: [valueRef],
+        errMsg: `${valueName} must be a real date`,
+      };
+    }
+    if (dayFormatValidation.error) {
+      return {
+        errRefs: [dayFieldRef],
+        errMsg: `${valueName} must be a real date`,
+      };
+    }
+    if (monthFormatValidation.error)
+      return {
+        errRefs: [monthFieldRef],
+        errMsg: `${valueName} must be a real date`,
+      };
+  }
+  if (yearFormatValidation.error) {
+    return {
+      errRefs: [yearFieldRef],
+      errMsg: `${valueName} must be a real date`,
+    };
+  }
+  try {
+    set(new Date(), {
+      year: Number(inputtedYear),
+      month: Number(inputtedMonth) - 1,
+      date: Number(inputtedDay),
+    });
+  } catch {
+    return {
+      errRefs: [valueRef],
+      errMsg: `${valueName} must be a real date`,
+    };
```

Comment:
I think `set(new Date(), { year: 2024, month: 1, date: 31})` will give the second of March. There is an example of one way to do this that treats 31st of Feb as an error here: https://github.com/UK-Export-Finance/dtfs2/blob/d8a0f1eac5d767e037711ea720e02922ca735dca/trade-finance-manager-ui/server/controllers/utilisation-reports/helpers/validate-payment-form-values.ts#L83-L90 https://github.com/UK-Export-Finance/dtfs2/blob/d8a0f1eac5d767e037711ea720e02922ca735dca/trade-finance-manager-ui/server/controllers/utilisation-reports/helpers/validate-payment-form-values.ts#L116-L119

Reply:
That was super helpful, thank you! Now added this in.